src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala

```diff
@@ -199,44 +199,75 @@ private[internal] trait TypeConstraints {
         //Console.println("solveOne0(tv, tp, v, b)="+(tvar, tparam, variance, bound))
         var cyclic = bound contains tparam
+        // This is attempting to reorder the solving of TypeVars such that tvs which are
+        // mentioned in the active bound are solved first.
+        // 1. Is this reordering desirable? Should solving always be left to right instead?
+        // 2. Should this take account of the variance of tparam2 and flip the bounds if
+        //    necessary?
         foreach3(tvars, tparams, variances)((tvar2, tparam2, variance2) => {
-          val ok = (tparam2 != tparam) && (
-            (bound contains tparam2)
-              ||  up && (tparam2.info.bounds.lo =:= tparam.tpeHK)
-              || !up && (tparam2.info.bounds.hi =:= tparam.tpeHK)
-            )
+          val ok = (tparam2 != tparam) && ((bound contains tparam2) || {
+            val bound2 = if (up) tparam2.info.bounds.lo else tparam2.info.bounds.hi
+            bound2 contains tparam
+          })
           if (ok) {
             if (tvar2.constr.inst eq null) cyclic = true
             solveOne(tvar2, tparam2, variance2)
           }
         })
+
         if (!cyclic) {
+          def propagateBounds(tparam2: Symbol): Unit = {
+            def addBound(variance: Variance, bound: Type) = if (bound != NoType) {
+              debuglog(s"$tvar.addBound($variance, $bound.instantiateTypeParams($tparams, $tvars))")
+              if (variance.isInvariant || variance.isCovariant) tvar addHiBound bound
+              if (variance.isInvariant || variance.isContravariant) tvar addLoBound bound
+            }
+
+            def toInst(tp: Type): Type = tp match {
+              case tv: TypeVar =>
+                val inst = tv.inst
+                if (inst == null) NoType else inst
+              case _ => tp
+            }
+
+            // Try harder to propagate information we can discover from this type variable's
+            // occurence in bounds on other type variables. Also avoid some ill-kinding issues.
+            def propagate(variance: Variance, bound: Type, instantiate: => Type): Unit = bound match {
+              // Propagate bounds of the same kind: F[A, B, C[_]] <: G[A, B, C] => F <: G
+              // Undoing eta-expansion to ensure the propagated bound is correct.
+              case GenPolyType(tparams, TypeRef(_, `tparam`, targs))
+                if tparams.corresponds(targs) { (tp, ta) => ta.typeArgs.isEmpty && ta.typeSymbolDirect == tp }
+                => addBound(variance, instantiate)
+              // Propagate type constructor: F[A] <: B => F <: B.typeConstructor
+              // Propagate type arguments: F[+A, -B] <: C => A <: C.typeArgs(0) && B >: C.typeArgs(1)
+              case TypeRef(_, tsym, targs) =>
+                val inst = toInst(instantiate)
+                if (tsym == tparam) addBound(variance, inst.typeConstructor)
+                foreach3(tsym.typeParams, targs, inst.typeArgs) {
```

Comment:
I'm not convinced of this part. Subtyping only moves to arguments once the type constructors are *equal*, not just in some subtype relation. You'd have to do something like relating F's type args with `C.baseType(F).typeArgs`, but we don't support doing that stuff on type vars. Can you show with an example that this code path is crucial, and how it works?

Reply:
In any case we have type variables on both sides so in the worst case we will infer something that doesn't respect the bounds and it will fail to compile when we check them. I couldn't find an example that compiled before and would fail with the more aggressive propagation, but if you can find one, that would be awesome.

Reply:
Thanks for bearing with me, this is tricky stuff and it's core to the type checker, so we need to take our time to understand the change and make sure it's right. 

I don't see how this step propagates the constraints to the right type args, since `addBound` will still be constraining the type var that was originally bounded by the applied type constructor, and now we are recursing on its args without changing the type var that `addBound` works on. 

Even if we did propagate those args to the right type var somehow, I'm not sure it would help. When evaluating the progress made in the test, it's important to keep in mind the actual types inferred using `-Xprint:typer`. In the covariant case (and equivalents) we just end up with `Nothing`.  In the example, this seems to be an accident, because the type args in bounds of CC end up constraining CC itself. (Turn the `debugLog` in `addBound` into a `println` to see the trace.)

Reply:
@adriaanm you're right, we _are_ propagating constraints to the type var currently under consideration. We're getting those constraints from bounds on _other_ type variables that it appears in.

For instance, consider `def foo[A <: List[B], B](x: A): Option[B] = None` where `A` is solved as `List[Int]` (see, `test/files/pos/bounds-propagation.scala`). Now when we come to solve for `B` we inspect the bounds that `B` appears in, here the bound on `A`. From that we can infer an `Int` bound on `B` and so solve `B` as `Int`. With the status quo we don't consider `B`'s occurrence in other bounds so the best we get is `Nothing`.

Apropos the use of `baseType` in this context, at least one of the participants will be an unsolved type var, so we'll not get anything useful out of it. As @joroKr21 says we'll be checking for within bounds later anyway, so I don't think there's a problem here.